concordium-consensus/src/Concordium/TransactionVerification.hs

```diff
@@ -0,0 +1,177 @@
+{-# LANGUAGE TypeFamilies #-}
+module Concordium.TransactionVerification
+  where
+
+import Control.Monad.Trans
+import Control.Monad.Trans.Reader
+import qualified Data.Map.Strict as OrdMap
+import qualified Data.HashMap.Strict as HM
+import qualified Data.Serialize as S
+
+import qualified Concordium.Types.Transactions as Tx
+import qualified Concordium.ID.AnonymityRevoker as AR
+import qualified Concordium.ID.IdentityProvider as IP
+import qualified Concordium.GlobalState.Types as GSTypes
+import qualified Concordium.GlobalState.BlockState as BS
+import qualified Concordium.Types.Parameters as Params
+import qualified Concordium.Types as Types
+import qualified Concordium.ID.Account as A
+import qualified Concordium.ID.Types as ID
+import Data.Maybe (isJust)
+import Control.Monad.Except
+import Concordium.Types.HashableTo (getHash)
+
+-- |The 'VerificationResult' type serves as an intermediate `result type` between the 'TxResult' and 'UpdateResult' types.
+-- VerificationResult's contains possible verification errors that may have occurred when verifying a 'AccountCreation' type.
+data VerificationResult
+  = Success
+  -- ^The verification passed
+  | Stale
+  -- ^The transaction was expired.
+  | ExpiryTooLate
+  -- ^The transaction had an expiry too distant in the future
+  | DuplicateAccountRegistrationID !ID.CredentialRegistrationID
+  -- ^The 'CredentialDeployment' contained an invalid registration id.
+  -- There already exists an account with the registration id.
+  | CredentialDeploymentInvalidIdentityProvider
+  -- ^The IdentityProvider does not exist for this 'CredentialDeployment'.
+  | CredentialDeploymentInvalidAnonymityRevokers
+  -- ^The anonymity revokers does not exist for this 'CredentialDeployment'.
+  | CredentialDeploymentInvalidKeys
+  -- ^The 'AccountCreation' contained invalid keys.
+  | CredentialDeploymentInvalidSignatures
+  -- ^The 'AccountCreation' contained invalid identity provider signatures.
+  | CredentialDeploymentExpired
+  -- ^The 'AccountCreation' contained an expired 'validTo'
```

Comment:
Such a verification result should not be cached, and should just immediately be rejected. 
VerificationResults that should be cached, hence they might be valid in the future can be seen here:
https://github.com/Concordium/concordium-node/blob/credential-verification/concordium-consensus/src/Concordium/TransactionVerification.hs#L50

Reply:
Yes the purpose of this is to provide the user with a meaningful error if the transaction of theirs failed. 

@abizjak Regarding "duplication checks" for a specific transaction. 
As I see, this will not happen because of the node's deduplication logic of incoming transactions, so I didn't see any reason to actually store the errorneus transaction verification result, as it will be immediatly rejected the first time it arrives and subsequently times the node will reject it before reaching consensus.

I admit this functionality is a bit implicit. 
The other option would be to implement the cache as e.g., an PSQ and remove expired ones as time goes on, then everything could be in the cache. But in reality they wouldn't be used as mentioned above.

Reply:
I would suggest having a separate type, e.g. `CacheableVerificationResult` that only has the checks that can be cached, to make states where other results are cached unrepresentable.

Reply:
Yes I think that is a good idea.